### ISSUE TITLE
d.ts with import statement exports types that are not marked exported

### ISSUE THREAD

**Issue by evmar** [Microsoft/TypeScript#19545]
**TypeScript Version:**  typescript@2.7.0-dev.20171027 (also 2.4 and 2.5)

**Code**

```ts
// extra.ts
export var x = 3;

// iface.d.ts
import {x} from './extra';
// note: this next line is *not* exported
declare interface Foo { bar: string; }

// user.ts
import {Foo} from './iface';
```

**Expected behavior:**
This should fail to compile, because `iface.d.ts` did not export the `Foo` type.  Error message should be
```
 error TS2305: Module ... has no exported member 'Foo'.
```

**Actual behavior:**
It compiles.

Extra notes:
1) if iface.d.ts has no import statement, but instead uses `export {}` to become a module, then expected error is produced
2) if you remove import/export statements from iface.d.ts, then as expected you get the 'is not a module' error
3) if you rename iface.d.ts to iface.ts, then expected error is produced

**Comment by andy-ms** [comment]
In a `.d.ts` file we normally assume that the things you're writing are public exports, else you would not be writing them.

> if iface.d.ts has no import statement, but instead uses export {} to become a module, then expected error is produced

This is interesting though, I didn't know `export {}` would make other things not exported. @RyanCavanaugh is that intentional?

**Comment by mhegazy** [comment]
Sometime between 0.8.3 and 0.9.1 we saw that ample `declare export` did not look that nice in .d.ts files. as a result `declare` was banned inside a `declare` block, and `export` was deemed "optional" in an ambient context (i.e. a `declare` declaration, or a .d.ts file).

ES6 modules come along, and we need to support them. `export {}` and `export default` have semantics that allow for exporting/renaming a local declaration. the current rules we had for .d.ts files did not fit with that. Just keeping the rule of implicit exporting means that we export declarations on the .d.ts that were not in the .ts. We added an exception for `export {}` and `export default`. 

If the .d.ts file/ambient module contains at least one `export {}` or an `export default`, only `export`ed declarations will be visible.
